File: 11.container-with-most-water.cpp
/*
 * @lc app=leetcode id=11 lang=cpp
 *
 * [11] Container With Most Water
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution
{
public:
    int maxArea(vector<int> &height)
    {
        int lo = 0, hi = height.size() - 1, max_area = 0;

        while (lo < hi)
        {
            int new_area = min(height[lo], height[hi]) * (hi - lo);
            max_area = max(max_area, new_area);
            if (height[lo] <= height[hi])
            {
                lo++;
            }
            else
            {
                hi--;
            }
        }

        return max_area;
    }
};
// @lc code=end

int main()
{
    Solution s;
    vector<int> input = {1, 3, 2, 5, 25, 24, 5};
    int output = s.maxArea(input);

    return 0;
}

File: 11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python3
#
# [11] Container With Most Water
#
from typing import *
# @lc code=start
class Solution:
    def maxArea(self, height: List[int]) -> int:
        p, q = 0, len(height) - 1
        
        max_area = 0
        while p < q:
            area = min(height[p], height[q]) * (q - p)
            max_area = max(max_area, area)
            
            if height[p] < height[q]:
                p += 1
            elif height[p] > height[q]:
                q -= 1
            elif height[p + 1] > height[q - 1]:
                p += 1
            elif height[p + 1] < height[q - 1]:
                q -= 1
            else:
                p += 1
        
        return max_area
            
# @lc code=end

if __name__ == '__main__':
    solution = Solution()
    a = solution.maxArea([1,3,2,5,25,24,5])
    print(a)

